src/codegen.c

```diff
@@ -8,73 +8,79 @@
 #include "jou_compiler.h"
 #include "util.h"
 
-struct LocalVariable {
-    char name[100];
-    LLVMValueRef pointer;
-};
+static LLVMTypeRef codegen_type(const struct Type *type)
+{
+    switch(type->kind) {
+    case TYPE_POINTER:
+        return LLVMPointerType(codegen_type(type->data.valuetype), 0);
+    case TYPE_SIGNED_INTEGER:
+    case TYPE_UNSIGNED_INTEGER:
+        return LLVMIntType(type->data.width_in_bits);
+    case TYPE_BOOL:
+        return LLVMInt1Type();
+    case TYPE_UNKNOWN:
+        assert(0);
+    }
+    assert(0);
+}
 
 struct State {
     LLVMModuleRef module;
     LLVMBuilderRef builder;
     const struct AstFunctionDef *ast_func;
     LLVMValueRef llvm_func;
+    // All local variables are represented as pointers to stack space, even
+    // if they are never reassigned. LLVM will optimize the mess.
     LLVMValueRef *llvm_locals;
 };
 
-static LLVMValueRef get_pointer_to_local_var(const struct State *st, const char *name)
+static LLVMValueRef get_pointer_to_local_var(const struct State *st, const struct CfVariable *cfvar)
 {
-    for (int i = 0; st->ast_func->locals[i].name[0]; i++)
-        if (!strcmp(st->ast_func->locals[i].name, name))
+    assert(cfvar);
+    for (int i = 0; i < st->ast_func->cfg->variables.len; i++) {
+        if (st->ast_func->cfg->variables.ptr[i] == cfvar) {
             return st->llvm_locals[i];
```

Comment:
This function is stupid. Maybe should calculate number of variables needed beforehand, so that I don't need to worry about growing the list and messing up old pointers? (That's why its a list of individually alloced pointers now)